core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java

```diff
@@ -364,18 +362,37 @@ boolean forceRefresh() {
             executor,
             metadataFuture,
             sslContextFuture);
-    refreshFuture.addListener(
-        () -> {
-          synchronized (instanceDataGuard) {
-            // update currentInstanceData with the most recent results
-            currentInstanceData = refreshFuture;
-            // schedule a replacement before the SSLContext expires;
-            nextInstanceData = executor
-                .schedule(this::performRefresh, secondsUntilRefresh(),
-                    TimeUnit.SECONDS);
+    Futures.addCallback(refreshFuture,
+        new FutureCallback<InstanceData>() {
+          public void onSuccess(InstanceData instanceData) {
+            synchronized (instanceDataGuard) {
+              // update currentInstanceData with the most recent results
+              currentInstanceData = refreshFuture;
+              // schedule a replacement before the SSLContext expires;
+              nextInstanceData = executor
+                  .schedule(() -> performRefresh(),
+                      secondsUntilRefresh(),
+                      TimeUnit.SECONDS);
+            }
           }
-        },
-        executor);
+
+          public void onFailure(Throwable t) {
+            logger.log(Level.WARNING,
+                "An error occurred while performing refresh. Retrying immediately.", t);
+            if (getInstanceData().getExpiration().toInstant().isBefore(Instant.now())) {
+              synchronized (instanceDataGuard) {
+                currentInstanceData = refreshFuture;
+                nextInstanceData = Futures.immediateFuture(performRefresh());
+              }
+            } else {
+              forceRefresh();
+            }
```

Comment:
I think the logic we want here is:
On error:
     if current is invalid:
          replace current
     next = new new refresh

Since we know this refresh failed, we always want to schedule a new one. We just don't want it to replace "current" if we think "current' might still be good.